src/main/java/com/basho/riak/client/api/commands/ChunkedResponseIterator.java

```diff
@@ -0,0 +1,138 @@
+    private final Function<ChunkT, BinaryValue> getContinuationFn;
+
+    private Iterator<CoreT> currentIterator = null;
+
+    public ChunkedResponseIterator(StreamingRiakFuture<ChunkT, ?> coreFuture,
+                                   int pollTimeout,
+                                   Function<CoreT, FinalT> createNextFn,
+                                   Function<ChunkT, Iterator<CoreT>> getNextIteratorFn)
+    {
+        this(coreFuture, pollTimeout, createNextFn, getNextIteratorFn, (x) -> null);
+    }
+
+    public ChunkedResponseIterator(StreamingRiakFuture<ChunkT, ?> coreFuture,
+                                   int pollTimeout,
+                                   Function<CoreT, FinalT> createNextFn,
+                                   Function<ChunkT, Iterator<CoreT>> getNextIteratorFn,
+                                   Function<ChunkT, BinaryValue> getContinuationFn)
+    {
+        this.timeout = pollTimeout;
+        this.coreFuture = coreFuture;
+        this.chunkQueue = coreFuture.getResultsQueue();
+        this.createNext = createNextFn;
+        this.getNextIterator = getNextIteratorFn;
+        this.getContinuationFn = getContinuationFn;
+        loadNextChunkIterator();
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+        if (currentIteratorHasNext())
+        {
+            return true;
+        }
+
+        loadNextChunkIterator();
+
+        return currentIteratorHasNext();
+    }
+
+    private boolean currentIteratorHasNext()
+    {
+        return currentIterator != null && currentIterator.hasNext();
+    }
+
+    private boolean possibleChunksRemaining()
+    {
+        // Chunks may remain if :
+        // Core Operation Not Done OR items still in chunk Queue
+        return !coreFuture.isDone() || !chunkQueue.isEmpty();
+    }
+
+    public boolean hasContinuation()
+    {
+        return continuation != null || possibleChunksRemaining();
+    }
+
+    public BinaryValue getContinuation()
+    {
+        return continuation;
+    }
+
+    @Override
+    public synchronized FinalT next()
+    {
+        return createNext.apply(currentIterator.next());
+    }
+
+    private void loadNextChunkIterator()
+    {
+        this.currentIterator = null;
+        boolean populatedChunkLoaded = false;
+
+        try
+        {
+            while (!populatedChunkLoaded && possibleChunksRemaining())
+            {
+                final ChunkT nextChunk = chunkQueue.poll(timeout, TimeUnit.MILLISECONDS);
+
+                if (nextChunk != null)
+                {
+                    this.currentIterator = getNextIterator.apply(nextChunk);
+                    populatedChunkLoaded = currentIteratorHasNext();
+
+                    loadContinuation(nextChunk);
+                }
+            }
+        }
+        catch (InterruptedException e)
+        {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
```

Comment:
My initial thought was exactly the same: set the flag and move forward without raising any exception. If this call was made from the whatever execution thread the programmer should take  care about the proper thread interuption and at least we do not make it worse.

Please excuse me for my hesitation.